TTL for cleaning up Jobs after they finish

janetkuo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: https://github.com/kubernetes/features/issues/592

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64470 
For https://github.com/kubernetes/features/issues/592

**Special notes for your reviewer**: @kubernetes/sig-apps-pr-reviews 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add a TTL machenism to clean up Jobs after they finish.
```

janetkuo committed:
Add TTLAfterFinished alpha feature

janetkuo committed:
Add TTL GC controller

kow3ns commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,271 @@
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+type TTLAfterFinishedController struct {
```
You should describe the type. You should add a comment explaining what the controller does and the design choice to have the contorller live outside of JobController.

janetkuo replied:
Done

kow3ns commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,271 @@
+	return tc
+}
+
+func (tc *TTLAfterFinishedController) Run(workers int, stopCh <-chan struct{}) {
```
Needs a comment.

janetkuo replied:
Done

kow3ns commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return nil, err
+	}
+	// TODO: handle finish time that happens in the future
```
You should come up with a well defined behavior for when this occurs. Consider simply deferring destroying the victim after the finish time. It's conservative, but it can't do harm.

janetkuo replied:
Done. Because it happens in the future it'll only delay destorying the Job.

kow3ns reviewed (approved):
Generally approve. Needs e2e as well.

kow3ns commented:
/approve

janetkuo commented on pkg/apis/batch/types.go:
```diff
@@ -138,6 +138,16 @@ type JobSpec struct {
+	// finishes. This field is alpha-level and is only honored by servers that
+	// enable the TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
```
Jobs finishes once its status is either complete or failed. Combined complete + failed into finished for simplicity, based on user feedback from kubeflow.

janetkuo commented:
@tnozicka this TTL controller only needs to watch Jobs and Pods for now; Jobs and Pods have different ways of indicating that they "finished". I'd like to switch to dynamic client/informer when we extend this to custom resources.

janetkuo committed:
Validate Job .spec.ttlSecondsAfterFinished; clear it when feature disabled

1. If TTLAfterFinished feature is enabled, the value should be non-negative.
2. If TTLAfterFinished feature is disabled, the field value should not
be kept.

janetkuo committed:
Add e2e test for TTL after finished

janetkuo commented on pkg/features/kube_features.go:
```diff
@@ -344,6 +344,12 @@ const (
+	// alpha: v1.12
+	//
+	// Allow TTL controller to clean up Pods and Jobs after they finish.
+	TTLAfterFinished utilfeature.Feature = "TTLAfterFinished"
```
Ditto

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,271 @@
+	clock clock.Clock
+}
+
+func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *TTLAfterFinishedController {
```
Done

kow3ns commented:
/approve

kow3ns commented:
/lgtm

janetkuo commented:
/assign @deads2k @liggitt 

Need your approval. Thanks!

cheftako commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.runJobCleanupWorker, time.Second, stopCh)
```
Given that we have multiple workers it might be worth using JitterUntil with a small jitter factor. Also wonder if it would be worth turning off sliding. (Only if you turn on Jitter)

Alternately firing off the wait.UntilNoSliding  every 1/workers seconds might be nice, though a bit complex.

janetkuo replied:
I'd prefer leaving it as-is. Other controllers with multiple workers also use Until.

tnozicka replied:
I agree with leaving this as is. I don't see a use case for the jitter here. The `wait.Until` is here to recover from hard failures and to wait a bit in between starting again.

cheftako commented on pkg/apis/batch/types.go:
```diff
@@ -138,6 +138,16 @@ type JobSpec struct {
+	// ttlSecondsAfterFinished limits the lifetime of a Job that has finished
+	// execution (either Complete or Failed). If this field is set, once the Job
+	// finishes, it will be deleted after ttlSecondsAfterFinished expires. If
+	// this field is unset, the Job will not be cleaned up after it finishes. If
```
Might be worth mentioning that it honors Finalizers.

janetkuo replied:
Updated API docs.

cheftako commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,292 @@
+		return
+	}
+
+	tc.jobsToCleanup.AddAfter(key, after)
```
We may want to watch out. This will become a blocking call once there are a thousand jobs who have yet to reach after. For small values of after I do not expect that to be a problem but we may want to consider what happens for large values of after and what thread is being blocked.

janetkuo replied:
Added a TODO for now.

janetkuo replied:
@cheftako code updated to use a heap to keep track of those jobs, and remove the use of AddAfter

janetkuo committed:
Make number of workers configurable

janetkuo commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -407,3 +408,14 @@ func startPVProtectionController(ctx ControllerContext) (http.Handler, bool, err
+	go ttlafterfinished.New(
+		ctx.InformerFactory.Batch().V1().Jobs(),
+		ctx.ClientBuilder.ClientOrDie("ttl-after-finished-controller"),
+	).Run(5, ctx.Stop)
```
Done.

janetkuo commented:
/test pull-kubernetes-integration

tnozicka commented:
Since I am starting to see approvals and lgtm here I am still missing KEP merged into community repository, asked for this 15 day ago - no response there (https://github.com/kubernetes/features/issues/592#issuecomment-409542217).

Also the FIXUP commits need to be cleaned up first.
/hold

I'll go through the code as well, just the initial feedback.

janetkuo commented:
FIXUP is just for easier review will be cleaned up before merging.

liggitt commented:
+1 on getting the community pull opened/merged. Did we decide against distinct ttls for success/failure cases?

liggitt commented on pkg/apis/batch/types.go:
```diff
@@ -138,6 +138,18 @@ type JobSpec struct {
+	// This field is alpha-level and is only honored by servers that enable the
+	// TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
```
status uses `CompletionTime`, should this be `TTLAfterCompletionSeconds`?

liggitt replied:
or is CompletionTime not set in failure cases?

janetkuo replied:
CompletionTime isn't set in failure cases.

janetkuo committed:
Autogen API docs

cheftako commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -146,6 +146,9 @@ func SetDefaults_KubeControllerManagerConfiguration(obj *KubeControllerManagerCo
 	if obj.HPAController.HorizontalPodAutoscalerUseRESTClients == nil {
 		obj.HPAController.HorizontalPodAutoscalerUseRESTClients = utilpointer.BoolPtr(true)
 	}
+	if obj.TTLAfterFinishedController.ConcurrentTTLSyncs == 0 {
```
<= 0?

You have this as an int not a uint. So we should handle the negative case.

janetkuo replied:
Done.

cheftako commented on pkg/controller/ttlafterfinished/util.go:
```diff
@@ -0,0 +1,59 @@
+func soonerExpiringJob(finishedJob1, finishedJob2 interface{}) bool {
+	j1, ok := finishedJob1.(*batch.Job)
+	if !ok {
+		return true
```
I believe this is a bug. It should be true that soonerExpiringJob(f1, f2) != soonerExpiringJob(f2, f1) as long as f1 != f2. This statement will violate that. (i.e. if both j1 and j2 are not batch jobs you should still be consistent about their sort order. You could use memory address or something else to achieve that.

janetkuo replied:
> It should be true that soonerExpiringJob(f1, f2) != soonerExpiringJob(f2, f1) as long as f1 != f2.

Is it required? If `f1 != f2` and `f1` and `f2` expire at the same time, both should return `false`. In [sort/sort.go](https://golang.org/src/sort/sort.go) it's also implemented similarly:
```go 
func (p StringSlice) Less(i, j int) bool { return p[i] < p[j] }
```

I want the error cases (i.e. non-Job objects or unfinished objects) to be placed on the top of the heap so that it can be dropped earlier instead of occupying space. It doesn't matter which error-case object is popped before the other one.

cheftako replied:
So comparison could be achieved by reflect.ValueOf(f1).Pointer() < reflect.ValueOf(f2).Pointer().
We may just want to consider that f1 == f2 for purposes of this sort.
That having been said if both f1 and f2 are not batch jobs we should return false (ie they are equal not less) but right now you will return true.

janetkuo replied:
Updated the code to treat error cases are f1==f2.

cheftako commented:
/lgtm

janetkuo commented:
Squashed commits after getting LGTM.

cheftako commented:
/lgtm

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,441 @@
+	if job.DeletionTimestamp != nil {
+		return
+	}
+
```
Shouldn't we just directly filter unmanaged jobs here with  `TTLSecondsAfterFinished == nil`? Same goes for the other functions.

janetkuo replied:
It's done in `checkTTLAndReact`. Actually, it makes sense to move the DeletionTimestamp check there too. Updated comments to make it more clear.

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,441 @@
+	}
+	err := tc.checkTTLAndReact(job, addJobToCleanupFn)
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %v: %v", job.Name, err)
```
namespace? (applies globally)

janetkuo replied:
Done

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,441 @@
+
+func (tc *TTLAfterFinishedController) addJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Adding job %s", job.Name)
```
namespace?

janetkuo replied:
Done

tnozicka reviewed (commented):
(went though about half of the PR) Any reason we can't use the regular controller structure with workqueue/sync_loop and AddAfter? This seems a bit complicated but maybe there is a reason in the rest of the PR I just didn't made it there yet.

janetkuo commented:
> Any reason we can't use the regular controller structure with workqueue/sync_loop and AddAfter?

Workqueue/sync_loop pattern (i.e. tc.jobsToCleanup & tc.cleanupJob) is used. I use an in-memory priority queue to store jobs to delete later instead of using `AddAfter`, because `AddAfter` relies on a channel of size 1k. If we have many jobs that need to be deleted a while later, the channel could be filled up. I'll document this reason in the comment.

janetkuo commented on pkg/controller/ttlafterfinished/util.go:
```diff
@@ -0,0 +1,71 @@
+	if !ok2 {
+		return false
+	}
+	// Error case: Jobs without expiration time are considered 'smaller'.
```
Jobs with no expiration time shouldn't stay in the delay clean up queue (as we won't clean them up), so pop them first instead of letting them occupy the space. Updated comments to clarify it.

janetkuo commented on pkg/registry/batch/job/strategy_test.go:
```diff
@@ -36,6 +42,12 @@ func newBool(a bool) *bool {
 }
 
+func newInt32(i int32) *int32 {
+	r := new(int32)
```
Oops, fixed. Fixed the above `newBool` function too.

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller_test.go:
```diff
@@ -0,0 +1,125 @@
+func TestTimeLeft(t *testing.T) {
+	now := metav1.Now()
+
+	testCases := map[string]struct {
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller_test.go:
```diff
@@ -0,0 +1,125 @@
+		expectErr        bool
+		expectedTimeLeft *time.Duration
+	}{
+		"Job unfinished":                 {metav1.Time{}, metav1.Time{}, int32Ptr(100), &now.Time, true, nil},
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller_test.go:
```diff
@@ -0,0 +1,125 @@
+		failedTime       metav1.Time
+		ttl              *int32
+		since            *time.Time
+		expectErr        bool
```
Done

janetkuo commented on hack/.golint_failures:
```diff
@@ -121,6 +121,7 @@ pkg/controller/service
 pkg/controller/serviceaccount
 pkg/controller/statefulset
 pkg/controller/ttl
+pkg/controller/ttlafterfinished
```
Removed it and fixed golint failures.

janetkuo commented:
@tnozicka thanks, comments addressed, PTAL.
I put the changes in "FIXUP commits" for easier review - let me know if you prefer them to be squashed.

janetkuo commented:
KEP was merged.

/test pull-kubernetes-e2e-kops-aws

janetkuo commented:
`HEAD` was broken (won't pass hack/verify) and I had to fix it here after rebasing. Possibly related to #66916.

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+
+	// No need to check delay cleanup queue for now.
+	if checkAt == nil || checkAt.IsZero() || nowUTC.Before(checkAt.UTC()) {
+		return true
```
Is this just hot looping in that case?

janetkuo replied:
It's checking the time to clean up every 10s, and only continue when it's time to clean up.

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+
+	// The next time we should check the jobsDelayCleanupQueue.
+	nextCheckAtMutex sync.RWMutex
+	nextCheckAt      *time.Time
```
Am I missing a place where we sleep for the diff between now and nextCheckAt, or a change in the delay queue?

janetkuo replied:
We cannot sleep because `nextCheckAt` might become earlier (e.g. when a Job that will expire earlier is added).

tnozicka replied:
you would get woken up by the new item being added to the queue, in golang terms:
```
select {
nextDelay : = delayqueuetop.addTime - now
case <- time.After(nextDelay):
case <- queueUpdatedCh:
}

tnozicka reviewed (changes requested):
I think the 2 queue design is a better fit that AddAfter, although the channel size limit could be likely bypassed by wrapping the call in a goroutine, it wouldn't be good for longer duration.
(Nit: I personally find the naming of those queues and related functions a bit confusing due to high similarity, maybe we could come up with more distinguishable names.)

I am not comfortable with how much we do in informer handlers though. If I recall correctly those calls are blocking the whole shared informer and we should do as little as possible there - especially concerning is acquiring the cleanup queue Lock.

I think we should restructure the controller a bit. The informer handlers should only check if `Spec.TTLSecondsAfterFinished` is set and add it to the Workqueue. WorkqueueWorker would determine if the job should be dropped, cleaned up now, or it should wait and put it to the Delayqueue. DelayqueueWorker would process the Delayqueue at intervals given by the time remaining for the item at Delayqueue, or be woken up by a signal that the Delayqueue has been updated and rechecked.

I don't feels strong if we end up doing the Job completion check in the informer handler or in the WorkqueueWorker, both has its advantages - you either save a workqueue addition or have a common place  to check it in the Worker. I guess for the Job itself that is better fitting in the informer handler; for the general case  if we were to extend the design to other resources and that check could include an API call e.g. to check status of a subobject, doing it in the Worker would be better.

tnozicka commented:
> @tnozicka thanks, comments addressed, PTAL.
I put the changes in "FIXUP commits" for easier review - let me know if you prefer them to be squashed.

Thanks, and apologies for the delay. I like those commits for easier review; my previous comment was only aimed at not tagging such PR yet so it doesn't merge prematurely.

tnozicka commented:
@janetkuo @kubernetes/sig-apps-pr-reviews maybe we should fix how AddAfter is implemented, I won't block this PR on that, but seems like other controllers have the same issue only the likelihood differs

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+	// Note: we cannot rely on workqueue.AddAfter, because it uses a buffered channel of
+	// limited size (1k), and the workqueue will be blocked after the channel is full.
+	jobsDelayCleanupQueueMutex sync.RWMutex
+	jobsDelayCleanupQueue      *cache.Heap
```
This is currently holding the whole Job object. Would it make sense for this queue to hold only something like
```golang
type Item struct {
    readyAt time.Time
    key string
}
```
which would make the concept generic and easily extensible for other resources (like Pods) we will add in future?

liggitt replied:
big +1 for only storing the key in the queue

janetkuo replied:
I'd prefer do this later when TTL controller is generalized to handle other resources. 

We currently watch Jobs and add Jobs to workqueue to process -- either clean up those Jobs or add them to the delay queue. We also pop Jobs from delay queue and add them to workqueue.

If we use `Item` in the delay queue, we will need to convert `Item` <--> `Job` when adding things to delay queue and when processing things in the workqueue. Since the workers of the workqueue only handle Jobs (i.e. get Jobs and delete Jobs), would it make sense to switch to something like `Item` later?

janetkuo replied:
I've updated it to store `Item` instead of `Job`.

tnozicka commented on pkg/registry/batch/job/strategy_test.go:
```diff
@@ -81,11 +94,16 @@ func TestJobStrategy(t *testing.T) {
 	if len(errs) != 0 {
 		t.Errorf("Unexpected error validating %v", errs)
 	}
+	if ttlEnabled != (job.Spec.TTLSecondsAfterFinished != nil) {
```
nit:

> If ttlEnabled is true, job.Spec.TTLSecondsAfterFinished != nil should also be true.

if `ttlEnabled is true` and `job.Spec.TTLSecondsAfterFinished is nil` (failing your condition) 
(it is enabled but not set)
the test says:
`"Job should only allow setting .spec.ttlSecondsAfterFinished when %v feature is enabled"`
which doesn't match.

janetkuo replied:
Done. Split 2 cases and error messages.

liggitt commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+		return nil
+	}
+
+	// Before deleting the Job, do a final sanity check just in case TTL is modified.
```
I don't really see what this buys us over using the version in our watch-fed cache... the ttl could still change between this lookup and the delete that follows

tnozicka replied:
I am in between, caches can go stale, but as you say we don't guarantee the atomicity either. On the other hand the cost isn't high here - you do only live get if you were to call delete. 

Maybe we should make that field immutable?

janetkuo replied:
@liggitt @tnozicka The benefit is that we can be sure that the TTL has expired at this point; we can't be sure with the value from cache. As long as the TTL expires at some point, it's safe for the system to clean up the resource.  

For example, 

1. At t1, we expect TTL to expire at t1+10
1. TTL was changed and it will expire at t1+20, but we haven't seen it in cache.
1. At t1+10, we check if TTL actually expires, and find out it hasn't. If we don't do this, we can't be sure if TTL has expired at any moment - if TTL was changed before t1+10, the new value should be respected. 
1. At t2+20 we check again, TTL expires and it's safe to clean up now. 
1. Even if TTL was changed afterwards, you can't change the fact that the TTL has expired at some point and the resource can be cleaned up.

tnozicka replied:
@janetkuo I think you are correct.

Is there a use case to support changing the TTL after you set it? 

Also I guess for pods the spec is immutable anyways hence the TTL field - do we want to differ with JobSpec?

janetkuo replied:
> Is there a use case to support changing the TTL after you set it?

Some example use cases:
* The users have some existing Jobs that already complete, and they want to set their TTLs 
* The users create some Jobs without TTLs, and decide to set the TTLs after the Jobs are created (can be done with mutating webhooks). 
* The users create some Jobs with TTLs, and decide to unset/extend TTLs before the Jobs complete or before the TTL expire.

Making TTLs mutable is convenient for the users without sacrificing correctness (we can still tell whether a TTL has expired or not). At the very least we should allow TTLs to be shortened, but IMO it's safe to allow users to extend/unset TTLs and that would make it more useful for the users. 

> for pods the spec is immutable anyways

Pod's `image` field is mutable, so it's okay to make pod TTL mutable too.

liggitt replied:
> The benefit is that we can be sure that the TTL has expired at this point; we can't be sure with the value from cache. As long as the TTL expires at some point, it's safe for the system to clean up the resource.

I can buy that, though it's worth documenting on the field that once the TTL has expired, the job can be deleted, even if an update to extend the TTL returns a successful API response.

Longer-term, we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

janetkuo replied:
> we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

This approach was discussed in the original proposal (in comments): https://goo.gl/YxtxTi
However the downside is that we'll have  a DELETE conflict whenever the resource (Job) is updated, but we actually only care about changes to TTL and nothing else. Moreover, DELETE doesn't return us the resource, so we'll need to GET the resource again (to get the latest TTL to re-add it to the queue) whenever there's a conflict.

tnozicka replied:
> However the downside is that we'll have a DELETE conflict whenever the resource (Job) is updated, but we actually only care about changes to TTL and nothing else. Moreover, DELETE doesn't return us the resource, so we'll need to GET the resource again (to get the latest TTL to re-add it to the queue) whenever there's a conflict.

As there is a conflict ("higher" RV) you can safely drop the item and wait for informers to catch up, no need for GET. I guess given the Job is suppose to be finished controllers don't really have much of a reason to update it and cause conflicts which leaves the person updating the TTL.

> Longer-term, we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

+1; I have needed that feature several times in the past

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+
+func timeLeft(j *batch.Job, since *time.Time) (*time.Duration, *time.Time, error) {
+	if !needsCleanup(j) {
+		return nil, nil, fmt.Errorf("Job %s/%s should not be cleaned up", j.Namespace, j.Name)
```
I'm just capitalizing "Job" (API object).

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+		}
+		job, ok := obj.(*batch.Job)
+		if !ok {
+			glog.V(2).Infof("Error peeking obj from the delay cleanup queue, it's not a Job, dropping it: %#v", obj)
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+	}
+	finishAtUTC := finishAt.UTC()
+	if finishAtUTC.After(sinceUTC) {
+		glog.V(4).Infof("Warning: Found Job %s/%s finished in the future. This is likely due to time skew in the cluster. Job cleanup will be deferred.", j.Namespace, j.Name)
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+		nowUTC = tc.clock.Now().UTC()
+		t, expireAt, err := timeLeft(job, &nowUTC)
+		if err != nil {
+			if jobutil.IsJobFinished(job) && job.Spec.TTLSecondsAfterFinished == nil {
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+				// This may happen when the user clear the TTL of a finished Job
+				glog.V(4).Infof("Detected Job %s/%s without TTL seconds, will not clean it up: %v", job.Namespace, job.Name, err)
+			} else {
+				glog.V(2).Infof("Unable to find the time to clean up Job %s/%s, dropping it: %v", job.Namespace, job.Name, err)
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+	}
+
+	tc.jobsDelayCleanupQueueMutex.Lock()
+	defer tc.jobsDelayCleanupQueueMutex.Unlock()
```
Moved it outside of informer handler

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+	}
+	// If we popped something other than what we want to pop, we need to add it back.
+	// We expect this not to happen.
+	if job.Namespace != top.Namespace || job.Name != top.Name {
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+		glog.V(2).Infof("Error: popped obj from the delay cleanup queue, but it's not a Job: %#v", obj)
+	}
+	// If we popped something other than what we want to pop, we need to add it back.
+	// We expect this not to happen.
```
The queue was locked between peek and pop.

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+	// `deleteJobFn` cascade deletes the Job when TTL expires
+	deleteJobFn := func(j *batch.Job) error {
+		policy := metav1.DeletePropagationForeground
+		return tc.client.BatchV1().Jobs(j.Namespace).Delete(j.Name, &metav1.DeleteOptions{PropagationPolicy: &policy})
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,446 @@
+
+// needsCleanup checks whether a Job has finished and has a TTL set.
+func needsCleanup(j *batch.Job) bool {
+	return jobutil.IsJobFinished(j) && j.Spec.TTLSecondsAfterFinished != nil
```
Done

janetkuo commented:
@tnozicka I believe I've addressed all comments, including restructure controller, store item (key+time) in delay queue, and rename some functions. PTAL.

janetkuo commented:
/test pull-kubernetes-bazel-build

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,478 @@
+			tc.popJobsDelayCleanupQueue(top)
+			continue
+		}
+		if top.ttl == nil {
```
shouldn't you get the value present value from an informer? I'd not store it in the Item struct as it can can more outdated

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,478 @@
+// if the TTL will expire later.
+func (tc *Controller) processTTL(job *batch.Job) (expired bool, err error) {
+	// We don't care about the Jobs that are going to be deleted, or the ones that don't need clean up.
+	if job.DeletionTimestamp != nil || !needsCleanup(job) {
```
`job.DeletionTimestamp != nil` this is already checked by the caller

tnozicka replied:
disregard, the live get it

janetkuo commented:
@tnozicka as discussed offline, switched back to `AddAfter` instead of implementing our own delay queue. PTAL.

tnozicka reviewed (commented):
nits, but new version with small informer handlers and using AddAfter looks much better and simpler - thanks @janetkuo for changing that!

After looking at AddAfter in detail it has turned out that it doesn't pose an issue for this design or in general.

/approve

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,478 @@
+	}
+	// Use the latest Job TTL to see if the TTL truly expires.
+	expired, err = tc.processTTL(fresh)
+	if err != nil || !expired {
```
Split the case

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,478 @@
+
+	// The Job's TTL has expired, but the Job TTL might be stale.
+	// Before deleting the Job, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
```
Done

janetkuo commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,478 @@
+		if (c.Type == batch.JobComplete || c.Type == batch.JobFailed) && c.Status == v1.ConditionTrue {
+			finishAt := c.LastTransitionTime
+			if finishAt.IsZero() {
+				// TODO: start our own timer
```
Removed comment

janetkuo commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -435,3 +438,19 @@ type ServiceControllerConfiguration struct {
+	// "kube-scheduler" is the default scheduler lock object name
+	SchedulerDefaultLockObjectName = "kube-scheduler"
+
+	SchedulerDefaultProviderName = "DefaultProvider"
```
Remove this; accidentally included this while rebasing

janetkuo commented:
Squashed commits

dixudx reviewed (approved):
/lgtm

janetkuo commented:
Rebased. Re-applying tag.

tnozicka reviewed (approved):
/lgtm

liggitt commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -378,6 +378,7 @@ func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc
 	controllers["clusterrole-aggregation"] = startClusterRoleAggregrationController
 	controllers["pvc-protection"] = startPVCProtectionController
 	controllers["pv-protection"] = startPVProtectionController
+	controllers["ttl-after-finished-controller"] = startTTLAfterFinishedController
```
while this is alpha, ensure it is in the ControllersDisabledByDefault list

janetkuo replied:
`startTTLAfterFinishedController` won't be started if TTLAfterFinished feature is disabled.

liggitt replied:
ah, missed that. that's fine.

liggitt commented on pkg/apis/batch/types.go:
```diff
@@ -138,6 +138,18 @@ type JobSpec struct {
+
+	// ttlSecondsAfterFinished limits the lifetime of a Job that has finished
+	// execution (either Complete or Failed). If this field is set, once the Job
+	// finishes, it will be deleted after ttlSecondsAfterFinished expires. When
```
might be worth nuancing "it will be deleted after ttlSecondsAfterFinished expires" into "`ttlSecondsAfterFinished` after the Job finishes, it is eligible to be automatically deleted" since there's no guarantee the deletion happens exactly at that moment.

janetkuo replied:
Updated

liggitt commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,309 @@
+		return nil, err
+	}
+	if finishAt.UTC().After(since.UTC()) {
+		glog.Warningf("Warning: Found Job %s/%s finished in the future. This is likely due to time skew in the cluster. Job cleanup will be deferred.", j.Namespace, j.Name)
```
I'm not seeing where this deferral is done. am I missing it?

liggitt commented:
one nit on the field doc, one comment about not auto-enabling the controller while in alpha, lgtm otherwise

janetkuo committed:
Autogen

make clean && make generated_files

janetkuo commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -340,6 +340,15 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
+		addControllerRole(&controllerRoles, &controllerRoleBindings, rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "ttl-after-finished-controller"},
+			Rules: []rbacv1.PolicyRule{
+				rbacv1helpers.NewRule("get", "list", "watch", "update", "delete").Groups(batchGroup).Resources("jobs").RuleOrDie(),
```
Removed "update"

janetkuo commented:
@liggitt thanks! Rebased and addressed comments.

liggitt commented:
API and RBAC change lgtm

/approve

liggitt commented:
readding lgtm as well

/lgtm

janetkuo commented:
Just fixing verify failure. Re-applying tag.

mikedanese commented on pkg/apis/batch/validation/validation_test.go:
```diff
@@ -213,6 +215,20 @@ func TestValidateJob(t *testing.T) {
 			},
 		},
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.TTLAfterFinished) {
```
Neither of these test cases are going to run since the feature won't be enable in these tests.

mikedanese commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -70,6 +76,11 @@ func (jobStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object
 	oldJob := old.(*batch.Job)
 	newJob.Status = oldJob.Status
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.TTLAfterFinished) {
```
Changes here are unnecessary, since they are caught in validation, right?

mikedanese commented on pkg/registry/batch/job/strategy_test.go:
```diff
@@ -24,16 +24,24 @@ import (
+)
+
+var (
+	ttlEnabled = utilfeature.DefaultFeatureGate.Enabled(features.TTLAfterFinished)
```
declare this in TestJobsStrategy since it's the only place you are using it.

mikedanese commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -172,6 +173,9 @@ func NewKubeControllerManagerOptions() (*KubeControllerManagerOptions, error) {
 			ConcurrentServiceSyncs: componentConfig.ServiceController.ConcurrentServiceSyncs,
 		},
+		TTLAfterFinishedController: &TTLAfterFinishedControllerOptions{
+			ConcurrentTTLSyncs: componentConfig.TTLAfterFinishedController.ConcurrentTTLSyncs,
```
I'm a proponent of YAGNI when it comes to configuration like parallel syncs. Is there a reason we want it explicitly for this feature as compared to other loops? An operator guessed value might be slighter better than a developer guessed value in an edge situation but the cost is extra code and suportability complexity. This was @cheftako's request

mikedanese replied:
I'd prefer adding workqueue latency and handletime metrics and explore adaptive worker counts and hand knobs to operators as a last resort.

mikedanese commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -378,6 +378,7 @@ func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc
 	controllers["clusterrole-aggregation"] = startClusterRoleAggregrationController
 	controllers["pvc-protection"] = startPVCProtectionController
 	controllers["pv-protection"] = startPVProtectionController
+	controllers["ttl-after-finished-controller"] = startTTLAfterFinishedController
```
`controller` is implicit in the context. All other names omit the `-controller`.

mikedanese commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,308 @@
+	job := cur.(*batch.Job)
+	glog.V(4).Infof("Updating job %s/%s", job.Namespace, job.Name)
+
+	if job.DeletionTimestamp == nil && needsCleanup(job) {
```
why not move `job.DeletionTimestamp == nil` into needsCleanup?

mikedanese commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller.go:
```diff
@@ -0,0 +1,308 @@
+}
+
+func getFinishAndExpireTime(j *batch.Job) (*time.Time, *time.Time, error) {
+	if !needsCleanup(j) {
```
Call this once at the begining of processJob, then don't check it again in all these helpers?

mikedanese commented:
/approve

I'd like to see whether we can remove the flag before beta, but I won't block this PR.

mikedanese commented:
per previous lgtms
/lgtm

janetkuo commented:
/test pull-kubernetes-e2e-kops-aws

tnozicka commented on pkg/controller/ttlafterfinished/ttlafterfinished_controller_test.go:
```diff
@@ -0,0 +1,177 @@
+		failedTime       metav1.Time
+		ttl              *int32
+		since            *time.Time
+		expectErr        bool
```
It will be equal, if your test case represents the actual error, not a (sub)string. Job name is fixed in the test - I don't see a reason why test only part of the error and leave the other part and the type untested; it even simplifies the test logic. I don't feel strong if you choose not to follow that path.
